server/src/main/java/org/apache/druid/client/selector/TierSelectorStrategy.java

```diff
@@ -47,6 +49,22 @@
   List<QueryableDruidServer> pick(
       Int2ObjectRBTreeMap<Set<QueryableDruidServer>> prioritizedServers,
       DataSegment segment,
-      int numServersToPick
-  );
+      int numServersToPick);
+
+  @Nullable
+  default <T> QueryableDruidServer pick(Query<T> query,
+      Int2ObjectRBTreeMap<Set<QueryableDruidServer>> prioritizedServers,
+      DataSegment segment)
+  {
+    return pick(prioritizedServers, segment);
```

Comment:
If you want to keep old interfaces, I think the old ones should have a default implementation that calls this new interface with `null` query. Then we can remove the implementations of old interfaces in `AbstractTierSelectorStrategy`.

Reply:
This was done for backward compatibility - I was trying to make this change as minimal as possible on the first cut. We don't have any custom implementations here and I haven't seen any on a cursory look, but I don't want to presume. I think it's cleaner to remove the ones that don't take Query and having an interface of all defaults that call each other makes the abstract methods less obvious. What do you think? I'm happy to defer to your judgement.

Reply:
I don't have strong opinion here. The current one looks good to me.

Reply:
works for me, current one it is!